grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -354,6 +358,24 @@ private static JavaType getJavaTypeForWellKnownTypes(FieldDescriptor fd) {
             return messageType.getFields().get(0).getJavaType();
         }
 
+        // The messages of the following types can be sent only via HTTP body.
+        if (Struct.getDescriptor().getFullName().equals(fullName) ||
```

Comment:
Don't we need to deal with a recursive gRPC `message` defined by users?
For example:
```java
message Recursive {
  string value = 1;
  Recursive nested = 2;
}
```
The `Recursive` seems to trigger `StackOverflowError`. In my view, a generalized circular detector by managing visited nodes needs to be implemented rather than hard-coded on well-known types.

Reply:
Agreed. I'm also thinking about that but could not figure out yet. Do you have any idea?

Reply:
I thought that we can use an `HashSet` to tract visited types.
For example:
```diff
diff --git a/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java b/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
index d1ee23544..1b3445678 100644
--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
@@ -25,6 +25,7 @@ import static java.util.Objects.requireNonNull;
 
 import java.io.IOException;
 import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
@@ -162,7 +163,10 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                 final Route route = routeAndVariables.getKey();
                 final List<PathVariable> pathVariables = routeAndVariables.getValue();
 
-                final Map<String, Field> fields = buildFields(methodDesc.getInputType(), ImmutableList.of());
+                final HashSet<Descriptor> visiting = new HashSet<>();
+                final Descriptor inputType = methodDesc.getInputType();
+                visiting.add(inputType);
+                final Map<String, Field> fields = buildFields(inputType, ImmutableList.of(), visiting);
 
                 int order = 0;
                 builder.put(route, new TranscodingSpec(order++, httpRule, methodDefinition,
@@ -274,7 +278,8 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
         return new SimpleImmutableEntry<>(builder.build(), PathVariable.from(segments, pathMappingType));
     }
 
-    private static Map<String, Field> buildFields(Descriptor desc, List<String> parentNames) {
+    private static Map<String, Field> buildFields(Descriptor desc, List<String> parentNames,
+                                                  Set<Descriptor> visiting) {
         final StringJoiner namePrefixJoiner = new StringJoiner(".");
         parentNames.forEach(namePrefixJoiner::add);
         final String namePrefix = namePrefixJoiner.length() == 0 ? "" : namePrefixJoiner.toString() + '.';
@@ -297,10 +302,19 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                     break;
                 case MESSAGE:
                     @Nullable
-                    final JavaType wellKnownFieldType = getJavaTypeForWellKnownTypes(field);
-                    if (wellKnownFieldType != null) {
+                    JavaType javaType = getJavaTypeForWellKnownTypes(field);
+                    final Descriptor messageType = field.getMessageType();
+                    if (javaType == null) {
+                        if (visiting.contains(messageType)) {
+                            // A recursive type.
+                            javaType = JavaType.MESSAGE;
+                        } else {
+                            visiting.add(messageType);
+                        }
+                    }
+                    if (javaType != null) {
                         builder.put(namePrefix + field.getName(),
-                                    new Field(field, parentNames, wellKnownFieldType));
+                                    new Field(field, parentNames, javaType));
                         break;
                     }
 
@@ -327,7 +341,7 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                     builder.putAll(buildFields(typeDesc, ImmutableList.<String>builder()
                                                                       .addAll(parentNames)
                                                                       .add(field.getName())
-                                                                      .build()));
+                                                                      .build(), visiting));
                     break;
             }
         });
```

Reply:
In addition, a `map<?,?>` type seems to be counted as a well-known type using `FieldDescriptor.isMapField()`.
```proto
message MapRequest {
  map<string, string> value = 1;
}
```

Reply:
Addressed. :-)

Reply:
Is there any reason that you throw the exception when it finds the recursive type instead of using @ikhoon'
s approach?

Reply:
I just want to mark a root field as message type which is impossible to be resolved by HTTP parameters.
For example,  the current way will mark `value` as a message type, but @ikhoon 's suggestion will mark `value.nested` as  a mesage type. 

```
  rpc EchoRecursive(EchoRecursiveRequest) returns (EchoRecursiveResponse) {
    option (google.api.http) = {
      get: "/v1/echo/recursive"

      additional_bindings {
        post: "/v1/echo/recursive"
        body: "value"
      }
    };
  }

message EchoRecursiveRequest {
  Recursive value = 1;
}

message EchoRecursiveResponse {
  Recursive value = 1;
}

message Recursive {
  string value = 1;
  Recursive nested = 2;
}
```

And I thought using a single hashset would not handle the following case (I didn't check):

```
message EchoRecursiveRequest {
  NestedMessage nested1 = 1;
  NestedMessage nested2 = 2;
}

message NestedMessage {
  string value = 1;
}
```